src/physics/base/ParticleTrackView.i.hh

```diff
@@ -0,0 +1,193 @@
+/*!
+ * Construct from dynamic and static particle properties.
+ */
+CELER_FUNCTION
+ParticleTrackView::ParticleTrackView(const ParticleParamsPointers& params,
+                                     const ParticleStatePointers&  states,
+                                     ThreadId                      id)
+    : params_(params), state_(states.vars[id.get()])
+{
+    REQUIRE(id < states.vars.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Initialize the particle
+ */
+CELER_FUNCTION ParticleTrackView&
+ParticleTrackView::operator=(const Initializer_t& other)
+{
+    REQUIRE(other.def_id < params_.defs.size());
+    REQUIRE(other.energy > 0);
+    state_ = other;
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+// DYNAMIC PROPERTIES
+//---------------------------------------------------------------------------//
+/*!
+ * Unique particle type identifier.
+ */
+CELER_FUNCTION ParticleDefId ParticleTrackView::def_id() const
+{
+    return state_.def_id;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Kinetic energy [MeV].
+ */
+CELER_FUNCTION real_type ParticleTrackView::energy() const
+{
+    return state_.energy;
+}
+
+//---------------------------------------------------------------------------//
+// STATIC PROPERTIES
+//---------------------------------------------------------------------------//
+/*!
+ * Rest mass [MeV / c^2].
+ */
+CELER_FUNCTION real_type ParticleTrackView::mass() const
+{
+    return this->particle_def().mass;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Elementary charge.
+ */
+CELER_FUNCTION real_type ParticleTrackView::charge() const
+{
+    return this->particle_def().charge;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Decay constant.
+ */
+CELER_FUNCTION real_type ParticleTrackView::decay_constant() const
+{
+    return this->particle_def().decay_constant;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Speed [1/c].
+ *
+ * Speed is calculated using the equality pc/E = v/c --> v = pc^2/E. Using
+ * \f[
+ * E^2 = p^2 c^2 + m^2 c^4
+ * \f]
+ * and
+ * \f[
+ * E = K + mc^2
+ * \f]
+ *
+ * the speed can be simplified to
+ * \f[
+ * v = c \sqrt{1 - (mc^2 / (T + mc^2))^2} = c \sqrt{1 - \gamma^{-2}}
+ * \f}
+ *
+ * By choosing not to divide out the mass, this expression will work for
+ * massless particles.
+ */
+CELER_FUNCTION real_type ParticleTrackView::speed() const
+{
+    // Rest mass energy
+    real_type mcsq = this->mass() * units::speed_of_light_sq;
+    // Inverse of lorentz factor (safe for m=0)
+    real_type inv_gamma = mcsq / (this->energy() + mcsq);
+
+    return units::speed_of_light * std::sqrt(1 - inv_gamma * inv_gamma);
```

Comment:
The Lorentz factor is only calculated for m > 0, since it yields infinity for massless particles (we have a REQUIRE(this->mass() > 0) in its method).

A simple way to circumvent this is by writing speed as a function of p (which is incompatible, but still defined for both massive and massless cases) and E (which encompasses both cases): v = pc^2/E.

Reply:
It's really just a computational change. The (safe for massless) code you had originally written was:
```c++
    // Calculating total energy: E = sqrt(p^c^2 + m^2c^4)
    real_type c2 = units::speed_of_light_sq;
    real_type c4 = c2 * c2;
    real_type p2 = this->momentum() * this->momentum();
    real_type m2 = this->mass() * this->mass();

    real_type E = std::sqrt(p2 * c2 + m2 * c4);

    // Returning v = pc^2/E
    return this->momentum() * c2 / E;
```
Each call to `momentum` there is a `sqrt`, which is quite expensive. I just algebraically eliminated it and simplified. I think it's good to do that where possible, especially for a core piece of code like the particle track view (the speed function is going to be called at least once per track per step).

I think from a physical point of view your answers make sense and could easily be added to the documentation. If the codebase weren't in a relatively unstable state, I would have (and initially did) made a separate commit to optimize the function that shows the original unit tests (which IMHO is the "real" proof of correctness) return numerically  the same value. I can revert the change and add that a separate  commit into the code history if you'd like.

Anyway, since this is the implementation of a function and the math expression is documented in its description, I  think that "mathematically correct, with a documented  derivation, numerically stable, and computationally efficient" is the way to go.

Reply:
And for what it's worth, in my eyes:
```c++
v = units::speed_of_light * std::sqrt(1 - inv_gamma * inv_gamma)
```
has an instantly readable "domain" for the result: the answer can *only* be between zero and the speed of light, even numerically. In the original answer it's not clear at all that's the case, and there's not even an  `ENSURE` to check that we don't accidentally have superluminal particle tracks ;)

Reply:
> Each call to momentum there is a sqrt, which is quite expensive
Ok, that is fair.

>  I can revert the change and add that a separate commit into the code history if you'd like.
Nah, things just need to work well and be documented. Which is basically what you wrote between quotes at the end of your comment.

> has an instantly readable range for the result: the answer can only be between zero and the speed of light
Yeah, I agree. To my eyes v = c(pc/E) is also very direct, but I am biased. The sqrt(1-x) does makes things obvious for anyone.

> there's not even an ENSURE to check that we don't accidentally have superluminal particle tracks.
I did not think of one, since the only way we could achieve that (given that our math is consistent) is with m < 0. I think we will have safeguards for constructing particles anyway.

Reply:
Yeah we will definitely have checks for nonnegative mass, but again it's not immediately obvious that there's not going to be a subtle numerical error that leaves us with c = `1 + 1e-15` or something, because there's a lot of floating point math being done on non-reduced algebraic expressions, and each operation has the potential to introduce an error.

Reply:
That a really good point.